src/util/ConfigManager/ConfigManager.cpp

```diff
@@ -0,0 +1,447 @@
+
+  // Add the location of the configuration option to the json.
+  keyToConfigurationOptionIndex_[ptr] = configurationOptions_.size();
+  // Add the configuration option.
+  configurationOptions_.push_back(std::move(option));
+}
+
+// ____________________________________________________________________________
+const std::vector<ConfigOption>& ConfigManager::getConfigurationOptions()
+    const {
+  return configurationOptions_;
+}
+
+// ____________________________________________________________________________
+const ConfigOption& ConfigManager::getConfigurationOptionByNestedKeys(
+    const VectorOfKeysForJson& keys) const {
+  // If there is an entry at the described location, this should point to the
+  // index number of the configuration option in `configurationOptions_`.
+  const nlohmann::json::json_pointer ptr{createJsonPointer(keys)};
+
+  if (keyToConfigurationOptionIndex_.contains(ptr)) {
+    return configurationOptions_.at(
+        keyToConfigurationOptionIndex_.at(ptr).get<size_t>());
+  } else {
+    throw NoConfigOptionFoundException(vectorOfKeysForJsonToString(keys),
+                                       printConfigurationDoc(true));
+  }
+}
+
+// ____________________________________________________________________________
+nlohmann::json ConfigManager::parseShortHand(
+    const std::string& shortHandString) {
+  // I use ANTLR expressions to parse the short hand.
+  antlr4::ANTLRInputStream input(shortHandString);
+  ConfigShorthandLexer lexer(&input);
+  antlr4::CommonTokenStream tokens(&lexer);
+  ConfigShorthandParser parser(&tokens);
+
+  // The default in ANTLR is to log all errors to the console and to continue
+  // the parsing. We need to turn parse errors into exceptions instead to
+  // propagate them to the user.
+  ThrowingErrorListener errorListener{};
+  parser.removeErrorListeners();
+  parser.addErrorListener(&errorListener);
+  lexer.removeErrorListeners();
+  lexer.addErrorListener(&errorListener);
+
+  // Get the top node. That is, the node of the first grammar rule.
+  ConfigShorthandParser::ShortHandStringContext* shortHandStringContext{
+      parser.shortHandString()};
+
+  // Walk through the parser tree and build the json equivalent out of the short
+  // hand.
+  return ToJsonConfigShorthandVisitor{}.visitShortHandString(
+      shortHandStringContext);
+}
+
+// ____________________________________________________________________________
+void ConfigManager::parseConfig(const nlohmann::json& j) {
+  // Anything else but a literal json object is not something, we want.
+  if (!j.is_object()) {
+    throw ConfigManagerParseConfigNotJsonObjectLiteralException(j);
+  }
+
+  /*
+  According to the documentation, when iterating over the entries of a
+  `nlohmann::json` object using `items()`, there can be error/problems, IF the
+  life time of the object, on which one called `items()` on, doesn't exceeds the
+  life time of the iteration. (See the documentation for more information.)
+
```

Comment:
I think I already wrote this: It is not a bug in `nlohmann::json`. It is the behavior in the C++ lifetime rules, that makes it tempting to write code that has undefined behavior, because temporaries don't live long enough.
In concrete (assume that `flatten` returns a new object, and `items()` is a function, that returns a reference into that object:
``` 
for (const auto& bla : j.flatten().items()) {
 //
}
```
In C++ up till C++20 the result of `flatten()` is immediately destroyed, and so `items()` is dangling. In C++23 it will live for the whole loop. I don't consider this a bug in `nlohmann::json`.

Reply:
Ah, thank you. I did not know that.